engine/client/secretprovider/vault.go

```diff
@@ -3,20 +3,44 @@ package secretprovider
 import (
 	"context"
 	"fmt"
+	"net/url"
 	"os"
 	"strings"
+	"time"
 
 	vault "github.com/hashicorp/vault/api"
 	auth "github.com/hashicorp/vault/api/auth/approle"
 )
 
+type dataWithTTL struct {
+	expiresAt time.Time
+	data      map[string]any
+}
+
 var (
 	vaultClient *vault.Client
-	vaultCache  = make(map[string]map[string]any)
+	vaultCache  = make(map[string]dataWithTTL)
```

Comment:
:thinking: just realized, I think this probably needs a lock right?

I don't think there's anything to stop parallel secret access? So we should make sure to lock when accessing this cache.

Reply:
Yep, seems to be the case since we're only synchronizing readers here: https://github.com/dagger/dagger/blob/31c82dcd8871ae0ced55f9e3b77eed207098c169/core/secret.go?plain=1#L215 which is where the provider gets called.